AdvLoggerPkg/Crates/RustAdvancedLoggerDxe/src/lib.rs

```diff
@@ -83,22 +101,14 @@ impl AdvancedLogger {
     }
 
     // initialize the AdvancedLogger by acquiring a pointer to the AdvancedLogger protocol.
-    fn init(&self, bs: *mut BootServices) {
-        assert!(!bs.is_null(), "BootServices should not be NULL");
-        let boot_services = unsafe { &mut ptr::read(bs) };
-
-        let mut ptr: *mut c_void = ptr::null_mut();
-
-        let status = (boot_services.locate_protocol)(
-            &ADVANCED_LOGGER_PROTOCOL_GUID as *const _ as *mut _,
-            ptr::null_mut(),
-            ptr::addr_of_mut!(ptr),
-        );
-
-        self.protocol.store(
-            if status == Status::SUCCESS { ptr as *mut AdvancedLoggerProtocol } else { ptr::null_mut() },
-            Ordering::SeqCst,
-        )
+    fn init(&self, boot_services_impl: &impl boot_services::BootServices) {
+        let protocol_ptr = match boot_services_impl.locate_protocol(&ADVANCED_LOGGER_PROTOCOL, None) {
+            Ok(Some(interface)) => interface as *mut AdvancedLoggerProtocolInterface,
+            Ok(None) => ptr::null_mut(),

```

Comment:
You can do

``` rust
let mut protocol_ptr = match boot_services_impl.locate_protocol(&ADVANCED_LOGGER_PROTOCOL, None) {
  Ok(Some(interface)) => interface as *mut AdvancedLoggerProtocolInterface,
  _ => ptr::null_mut(),
}
```

Reply:
Yeah, I liked the explicit handling slightly better, and it was only one more line.